Fixed ppp_lwip_disconnect hangs when connection failure.

jarvte opened the pull request:
### Description
- fixed ppp_lwip_disconnect hangs when connection failure.
- improved nsapi_ppp_connect error handling
- fixing issue https://github.com/ARMmbed/mbed-os/issues/6979

@AriParkkila @kjbracey-arm please review.

Internal ref to defect: IOTCELL-1028

### Pull request type
    [X] Fix
    [ ] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

Commit:
Fixed ppp disconnect hangs when connection failure.

jarvte commented:
@0xc0170 travis failure not related to this fix, probably ci/pr-head neither.

kjbracey commented on features/FEATURE_LWIP/lwip-interface/ppp_lwip.cpp:
```diff
@@ -295,20 +295,18 @@ extern "C" err_t ppp_lwip_connect(void *pcb)
-    my_stream = NULL;
+        /* close call made, now let's catch the response in the status callback */
+        sys_arch_sem_wait(&ppp_close_sem, 0);
+        ppp_active = false;
```
Probably should actually set ppp_active to false even if ppp_close returns an error. It's probably down if that happens.

jarvte replied:
true, fixed

cmonr commented:
@jarvte Thanks for that comment. Restarted that Travis job...

Commit:
Review fix:
- set ppp_active false if close fails in ppp disconnect.
- unset sigio in ppp disconnect
- take ownership of filehandle in CellularNetwork::disconnect even in case of failure

AriParkkila commented on features/FEATURE_LWIP/lwip-interface/ppp_lwip.cpp:
```diff
@@ -295,20 +295,19 @@ extern "C" err_t ppp_lwip_connect(void *pcb)
+    if (ppp_active) {
+        err_t ret = ppp_close(my_ppp_pcb, 0);
+        if (ret != ERR_OK) {
+            ppp_active = false;
```
All paths set `ppp_active = false;`, could be done once?

jarvte replied:
ppp_active can't be called before sys_arch_sem_wait but did some modifications.

AriParkkila commented on features/FEATURE_LWIP/lwip-interface/ppp_lwip.cpp:
```diff
@@ -373,6 +372,9 @@ nsapi_error_t nsapi_ppp_connect(FileHandle *stream, Callback<void(nsapi_event_t,
             my_interface = NULL;
+            my_stream->set_blocking(true);
+            my_stream = NULL;
+            connection_status_cb = NULL;
```
AT_CellularNetwork::open_data_channel should release file handle before calling nsapi_ppp_connect(), and also reinitialise file handle if nsapi_ppp_connect() returns failure.

kjbracey replied:
By "release" do you mean unset its `sigio` callback? Yes, probably wise.

jarvte replied:
no need to do it here, only after my_interface->bringup which will set sigio. Added unset of sigio there. Good point!

AriParkkila commented on features/FEATURE_LWIP/lwip-interface/ppp_lwip.cpp:
```diff
@@ -390,7 +398,19 @@ nsapi_error_t nsapi_ppp_connect(FileHandle *stream, Callback<void(nsapi_event_t,
+    my_stream->set_blocking(true);
+    my_stream = NULL;
+
+    return retcode;
```
AT_CellularNetwork::disconnect should probably take ownership also in case of if (err != NSAPI_ERROR_OK)

jarvte replied:
true, fixed to AT_Cellularnetwork::disconnect()

0xc0170 commented:
/morph build

0xc0170 commented:
/morph export-build
/morph mbed2-build

jarvte commented:
@0xc0170 please merge
